### PR TITLE
feat(gfx): compose the streaming pipeline core (P6, #4040)

### PR DESCRIPTION
**Pull request by zackees**
P6's stages have existed as separate modules for a while — decode, primaries→XYZ, chromatic adaptation, the device solve, the gamut mapper, the flux scalar — with **nothing composing them**. `linearRgbToXyzQ16` was referenced only by its own module. This is the composition.

One bind-time build, one per-pixel call, no state carried between calls and no allocation — so it runs inside `PixelController` iteration without an RGB16 framebuffer (C2).

## Order, and why

1. `decodeTransferU16` — RGB8 code straight to u16 linear light. One semantic conversion (B3).
2. `linearRgbToXyzQ16` — with chromatic adaptation **folded into the same matrix at bind time**, so it costs nothing per pixel.
3. The gamut mapper, which owns clamping — the stages before it hand over the honest target even when it's outside the hull.
4. `applyFluxScalar` last: C4's single amplitude stage, where scaling all drives together preserves chromaticity by construction.

## Conformance

Scored against the P5 reference on its own 48 `rgb` vectors, across all three source profiles:

**Worst drive error 0.001312** — about 86 raw units of s16.16. A1's luminance budget is 0.5%; this is **0.13%**.

The test pins both the meaningful threshold (under one code at 8-bit output, so no vector differs by something a device could show) *and* the measured value, so a regression that stays under a code still fails rather than sliding by.

## The B3 structural test

`ci/tests/test_no_rgb8_intermediate.py`, companion to the existing A3/B11 one. B3 is a shape, not a number: a pipeline that quantized to 8 bits mid-chain would still produce plausible colours, just worse ones, and the error would read as rounding rather than as a missing requirement. It includes a test that the check can actually fail, since it's all regexes over text.

## Two things the tests caught, both mine

**Full white doesn't come back as three full drives.** This profile's emitters are normalized to unit *luminance*, so reaching D65 needs roughly 0.21 : 0.72 : 0.07. The test now verifies the chromaticity the drives produce, rather than three numbers the solve would be comparing against itself.

**`buildSourceMatrixQ16` accepted primaries with red and green identical** — against its own documented contract (*"False if the primaries are collinear"*). Root cause is in `invert3x3`, whose singularity guard is `fabs(det) < 1e-20`. For chromaticity matrices with entries of order 1, float32's cancellation floor is ~1e-7: that exactly-singular set computes **−1.094e-7** and sails through, yielding an inverse scaled by `1/det ≈ −9.1e6` of pure rounding noise.

Guarded here geometrically instead — twice the primary triangle's area, a quantity with meaning at this scale (sRGB's is 0.224, against a 1e-4 threshold). The test covers both exact degeneracy *and* a near-degenerate set, since the exact case gives an area of precisely zero and so wouldn't exercise the threshold; the near case is the one that discriminates, confirmed by mutation.

The shared helper is **#4194**. This PR doesn't touch it — a relative determinant test changes behaviour for every caller and wants a maintainer's call.

Part of #4034 / #4040.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

- **New Features**
  - Added a Q16 streaming graphics pipeline that converts RGB8 pixels directly into emitter drive values.
  - Added support for configuring pipelines from source and device color profiles, adjusting brightness, and processing pixels without intermediate RGB8 buffers.
  - Added validation to reject profiles with degenerate color primaries.

- **Tests**
  - Added conformance coverage for color accuracy, brightness scaling, output bounds, edge cases, and invalid profiles.
  - Added safeguards against unintended RGB8 intermediate processing.

### CODE REVIEW

File: ci/tests/test_no_rgb8_intermediate.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+# A declaration of one of these on the per-pixel path is the RGB8 round trip
+# B3 forbids. `CRGB` and `CRGBW` are the 8-bit pixel types; a bare `u8`
+# array is the hand-rolled version of the same thing.
+FORBIDDEN_TYPES = ("CRGB", "CRGBW")
+
+# `u8 name[` -- an 8-bit buffer. Single `u8` scalars are fine and common
+# (a code coming in, a channel count), which is why this looks for the
+# array form specifically rather than for the type.
+U8_BUFFER = re.compile(r"\bu8\s+\w+\s*\[")
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_no_rgb8_intermediate.py]: Taken. `u8 \w+ \[` matched `u8 buffer[3];` but not `u8 code, intermediate[3];`, where `code` is a legitimate scalar and `intermediate` is exactly the buffer B3 forbids.

The scan now runs from `u8` to the first `[`, refusing to cross `;{}()` or `=`. The `;` stops one statement's `u8` matching the next statement's subscript; the `=` is what keeps `u8 value = table[i];` from reading as a declaration, since there the bracket is a subscript on the right-hand side.

Added to the self-test, along with the cases it must *not* fire on — a `u8` scalar, a right-hand-side subscript, function parameters, a 16-bit table. A structural check that flags ordinary code gets turned off, which fails the same way as one that catches nothing.

File: ci/tests/test_no_rgb8_intermediate.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    *discusses* RGB8, because explaining why there is no RGB8 intermediate is
+    exactly what their comments are for. The same reasoning as the sibling
+    test's -- a regex over C++ is a deliberate trade, and the cost of getting
+    it wrong here is a false alarm rather than a silently weakened guard,
+    because a stripped-out declaration would still have to appear somewhere
+    this test looks.
+    """
+
+    source = re.sub(r"/\*.*?\*/", " ", source, flags=re.S)
+    source = re.sub(r"//[^\n]*", " ", source)
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_no_rgb8_intermediate.py]: Confirmed and fixed. Stripping comments first treats the `//` in `"http://example.com"` as a line comment and eats the rest of the line — which would hide real code from the check, so it fails toward silence rather than toward noise.

Simply reversing the order does not work either: a quote inside a comment would then open a string literal and swallow everything to the next quote. So string literals, character literals, block comments and line comments are now matched in a single alternation, and precedence falls to whichever form starts first — which is what the language actually does.

Both directions are in the self-test now: a URL literal must leave the following code visible, and an apostrophe in a comment must not open a string.

**Comment by zackees** [comment]
Both findings taken and fixed; replies are on each thread. Both were tests that would have passed against the regression they were written to catch, and both fixes are checked by mutation.

@coderabbitai resolve

File: ci/tests/test_no_rgb8_intermediate.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+# delimiters, so precedence is decided by which starts first rather than by
+# the order of separate substitutions. Stripping comments before strings
+# would treat the `//` in a URL literal as a comment and eat the rest of the
+# line; stripping strings before comments would treat a quote inside a
+# comment as opening one.
+_LEXICAL = re.compile(
+    r'"(?:[^"\\\n]|\\.)*"'
+    r"|'(?:[^'\\\n]|\\.)*'"
+    r"|/\*.*?\*/"
+    r"|//[^\n]*",
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_no_rgb8_intermediate.py]: Taken. A raw literal can hold both a quote and a line-comment marker in one string, so whichever plain form matched it first would run off the end and swallow whatever followed — hiding real code from the check, which is the direction this guard must not fail in.

Handled by matching the raw form first, with the delimiter captured and back-referenced. That is what makes the terminator unambiguous, and it is the reason the form exists at all: `R"xy(a)" b)xy"` contains a `)"` in its body that is not the end of the string. Both shapes are pinned in the self-test.

The fix is load-bearing rather than defensive decoration — removing the raw-string alternative fails two of the six tests.

One note for the record: the sibling guard, `test_no_iterative_solver_per_pixel.py`, still documents raw strings as an accepted limit. I think that remains a reasonable call there — it looks for calls to three named symbols, where a missed match costs a false negative on a name nobody is hiding deliberately — but the two files now differ in rigour, and this one is the better lexer of the pair.
